Pointwise boundary surface values

Arpit-Babbar opened the pull request:
This PR add pointwise values on the surface like the coefficient of pressure and friction.

The computation has been verified using the reference data from [Roy Charles Swanson, Stefan Langer, Structured and Unstructured Grid Methods (2016)](https://ntrs.nasa.gov/citations/20160003623)

The verification requires plotting along with reference data, for which the script is available at https://gist.github.com/Arpit-Babbar/924823c08f4be962856a99c107b18fb5

TODO

- [x] Add tests for pointwise surface values (How to do it?)
- [x] The many TODOs within the code

Arpit-Babbar committed:
Add analysis_surface_2d.jl

Arpit-Babbar committed:
Add analysis_surface_2d.jl

Arpit-Babbar committed:
Format

Arpit-Babbar committed:
Format

Arpit-Babbar commented:
I currently have two issues

1) The output of pointwise values is saved into a `.txt` file. Does that need to be changed?

2) This is currently being computed as an `analysis_integral` quantity in the `AnalysisCallback`, but it is not an integral quantity. Where else can this belong? I think that it should be given its own place within the `AnalysisCallback`, one that does not print any values to the screen (because there are no values to be printed).

andrewwinters5000 reviewed (commented):
I did not do a full review, just left a question about the need to add another package dependency. As far as your other concerns here are my preliminary thoughts.

1. A `.txt` file is fine as ACSII data can be post-processed by nearly any program. However, if we are willing to use `DelimitedFiles` would something like `.csv` be better? Just thinking out loud.
2. Since no integral is computed, then, as far as I understand it, the `analysis_integral` is basically being overloaded for easier dispatch purposes. Then the pointwise values are computed on the fly with similar drag and lift functions and the write-to-file occurs within the loop over all nodes instead of afterwards. My initial reaction is that this could be confusing. It would be better to be more explicit and create a new generic function called something like `analysis_pointwise` that could still reuse the caches and such from the `AnalysisCallback` struct.

DanielDoehring commented on examples/p4est_2d_dgsem/elixir_navierstokes_NACA0012airfoil_mach08.jl:
```diff
@@ -142,14 +141,29 @@ lift_coefficient_shear_force = AnalysisSurfaceIntegral(semi, force_boundary_name
                                                            lift_coefficient,
                                                            drag_coefficient_shear_force,
-                                                           lift_coefficient_shear_force))
+                                                           lift_coefficient_shear_force,
+                                                           friction_coefficient,
+                                                           pressure_coefficient))
```
We should implement this in some fashion that this is not treated/labeled as an integral to avoid confusion.

DanielDoehring commented on src/callbacks_step/analysis_surface_2d.jl:
```diff
@@ -0,0 +1,277 @@
+#! format: noindent
+
+# This file contains callbacks that are performed on the surface like computation of
+# surface forces
```
```suggestion
# point-wise surface forces
```

DanielDoehring commented on src/callbacks_step/analysis_surface_2d.jl:
```diff
@@ -0,0 +1,277 @@
+                                                          boundary_symbol_or_boundary_symbols,
+                                                          variable)
+
+This struct is used to compute pointwise surface values of a quantity of interest `variable`
```
```suggestion
This struct is used to compute point-wise surface values of a quantity of interest `variable`
```

DanielDoehring commented on src/callbacks_step/analysis_surface_2d.jl:
```diff
@@ -0,0 +1,277 @@
+or `boundary_symbols`.
+For instance, this can be used to compute the surface pressure coefficient [`SurfacePressureCoefficient`](@ref) or
+surface friction coefficient [`SurfaceFrictionCoefficient`](@ref) of e.g. an airfoil with the boundary
+name `:Airfoil` in 2D.
```
```suggestion
symbol `:Airfoil` in 2D.
```

We should clarify this also here:

https://github.com/trixi-framework/Trixi.jl/blob/655497374d9991a7bd814b08af6751e81becab7d/src/callbacks_step/analysis_surface_integral_2d.jl#L21

Arpit-Babbar replied:
Doesn't it still say `name`?

DanielDoehring commented on src/callbacks_step/analysis_surface_2d.jl:
```diff
@@ -0,0 +1,277 @@
+- `linf::Real`: Reference length of geometry (e.g. airfoil chord length)
+"""
+function SurfacePressureCoefficient(pinf, rhoinf, uinf, linf)
+    psi = (zero(pinf), zero(pinf)) # Dummy value/placeholder, will not be used
```
I guess we should implement this without the need to set `psi`.

DanielDoehring commented on src/callbacks_step/analysis_surface_2d.jl:
```diff
@@ -0,0 +1,277 @@
+    n_elements = length(indices)
+
+    coord_array = Array{real(dg)}(undef, n_elements * n_nodes, dim)   # physical coordinates of indices
+    variable_array = Array{real(dg)}(undef, n_elements * n_nodes, 1) # variable values at indices
```
Why not make this a vector ?

DanielDoehring commented on src/callbacks_step/analysis_surface_2d.jl:
```diff
@@ -0,0 +1,277 @@
+    n_nodes = nnodes(dg)
+    n_elements = length(indices)
+
+    coord_array = Array{real(dg)}(undef, n_elements * n_nodes, dim)   # physical coordinates of indices
```
Probably more efficient making this a matrix and thus restricting to 2d ?

DanielDoehring commented on src/callbacks_step/analysis_surface_2d.jl:
```diff
@@ -0,0 +1,277 @@
+    n_nodes = nnodes(dg)
+    n_elements = length(indices)
+
+    coord_array = Array{real(dg)}(undef, n_elements * n_nodes, dim)   # physical coordinates of indices
+    variable_array = Array{real(dg)}(undef, n_elements * n_nodes, 1) # variable values at indices
```
Same as above

DanielDoehring commented:
I suggest we adapt/copy this code

https://github.com/trixi-framework/Trixi.jl/blob/655497374d9991a7bd814b08af6751e81becab7d/src/callbacks_step/analysis.jl#L583-L607

to non-integrated quantities, i.e., some which are not returning a scalar value and then call this here:

https://github.com/trixi-framework/Trixi.jl/blob/655497374d9991a7bd814b08af6751e81becab7d/src/callbacks_step/analysis.jl#L487-L488

DanielDoehring commented:
Regarding the type of file, I think we should aim for something that is easily post-processed, as now one does not get a single value but instead many, which are most likely also accompanied by spatial positions.

DanielDoehring committed:
Apply suggestions from code review

DanielDoehring committed:
Apply suggestions from code review

DanielDoehring committed:
Pointwise analysis

DanielDoehring committed:
Pointwise analysis

DanielDoehring committed:
Merge branch 'main' into surface_pressure_friction

DanielDoehring committed:
cosmetics

DanielDoehring committed:
cosmetics

DanielDoehring committed:
routines similar to timeseries

DanielDoehring committed:
routines similar to timeseries

DanielDoehring committed:
write pointwise data to jhdf5

DanielDoehring committed:
write pointwise data to jhdf5

DanielDoehring committed:
fmt

DanielDoehring committed:
fmt

DanielDoehring committed:
Merge branch 'main' into surface_pressure_friction

DanielDoehring committed:
not needed

DanielDoehring committed:
not needed

DanielDoehring committed:
rm unnecessary

DanielDoehring committed:
rm unnecessary

DanielDoehring committed:
re-organize

DanielDoehring committed:
re-organize

DanielDoehring commented:
The values are now written off in the usual fashion (as for e.g. restart & save solution) callbacks.

Please have a look @sloede @Arpit-Babbar

DanielDoehring commented on src/callbacks_step/analysis_surface_2d.jl:
```diff
@@ -0,0 +1,277 @@
+    t_trunc = @sprintf("%.3f", t)
+    filename = varname(variable) * t_trunc * ".txt"
+    # TODO - Should we start with a bigger array and avoid hcat?
+    writedlm(joinpath("out", filename), hcat(coord_array, variable_array))
```
We are most likely writing to `h5` file, thus I resolve this convo.

DanielDoehring committed:
remove todo

DanielDoehring committed:
remove todo

DanielDoehring committed:
Merge branch 'main' into surface_pressure_friction

Arpit-Babbar commented on src/callbacks_step/analysis.jl:
```diff
@@ -606,6 +619,26 @@ function analyze_integrals(analysis_integrals::Tuple{}, io, du, u, t, semi)
     nothing
 end
 
+# Iterate over tuples of analysis integrals in a type-stable way using "lispy tuple programming".
```
These are not integrals, right?

```suggestion
# Iterate over tuples of pointwise analysis quantities in a type-stable way using "lispy tuple programming".
```

DanielDoehring replied:
Yes, that is a copy-paste error.

Arpit-Babbar commented on src/callbacks_step/analysis.jl:
```diff
@@ -606,6 +619,26 @@ function analyze_integrals(analysis_integrals::Tuple{}, io, du, u, t, semi)
+function analyze_pointwise(analysis_quantities::NTuple{N, Any}, du, u, t,
+                           semi, iter) where {N}
+
+    # Extract the first analysis integral and process it; keep the remaining to be processed later
```
Same as above. These are not integrals.

```suggestion
    # Extract the first pointwise analysis quantity and process it; keep the remaining to be processed later
```

Arpit-Babbar commented on src/callbacks_step/analysis.jl:
```diff
@@ -606,6 +619,26 @@ function analyze_integrals(analysis_integrals::Tuple{}, io, du, u, t, semi)
+
+    analyze(quantity, du, u, t, semi, iter)
+
+    # Recursively call this method with the unprocessed integrals
```
Same as above. These are not integrals.

```suggestion
    # Recursively call this method with the unprocessed pointwise analysis quantities
```

Arpit-Babbar commented on src/callbacks_step/analysis_surface_2d.jl:
```diff
@@ -0,0 +1,290 @@
+name `:Airfoil` in 2D.
+
+- `semi::Semidiscretization`: Passed in to retrieve boundary condition information
+- `boundary_symbol_or_boundary_symbols::Symbol|Vector{Symbol}`: Name(s) of the boundary/boundaries
```
If we are changing the above to `symbol`, should we change this too?

```suggestion
- `boundary_symbol_or_boundary_symbols::Symbol|Vector{Symbol}`: Name(s) of the boundary/boundaries stored as symbol(s)
```

Arpit-Babbar reviewed (commented):
I really like the `h5` file solution. The design looks fine, although some TODO's and a proper review is still pending from my side.

DanielDoehring committed:
Apply suggestions from code review

Co-authored-by: Arpit Babbar <arpitbabbar@gmail.com>

DanielDoehring committed:
Apply suggestions from code review

Co-authored-by: Arpit Babbar <arpitbabbar@gmail.com>

DanielDoehring commented:
Regarding one of the TODOs: With mean value you are probably referring to the mean value of the quantity within that element, right? I see your point that one might be interested in computing this as well, as this cannot be easily recovered at post-processing.

At the same time, I have no clear use-case for this in mind.

Concerning sorting with increasing x value: This sounds reasonable to me, as this adds at least some structure to the output.

DanielDoehring committed:
Sort

DanielDoehring committed:
Sort

DanielDoehring committed:
remove prnt

DanielDoehring committed:
remove prnt

DanielDoehring committed:
Merge branch 'surface_pressure_friction' of https://github.com/Arpit-Babbar/Trixi.jl into ArpitBabbar_PointWiseSurfForces

DanielDoehring commented:
The data-export looks good to me. For some reason one needs to transpose the array when using `h5py`.

```python
import h5py
import numpy as np

# Open the file in read mode
with h5py.File('CF_x_000000.h5', 'r') as f:
  # List all groups
  print("Keys: %s" % f.keys())
  
  # Iterate over all keys
  for key in f.keys():
    data = f[key]

    data_array = np.array(data)
    if np.ndim(data_array) > 1:
      # Transpose to get dimensions right
      data_array = data_array.T

    print("Dimensions of array for key %s: %s" % (key, data_array.shape))
    print("Data for key %s: %s" % (key, data_array))
```

DanielDoehring committed:
name/symbol

DanielDoehring committed:
name/symbol

DanielDoehring committed:
Merge branch 'main' into surface_pressure_friction

sloede commented on src/callbacks_step/analysis_surface_2d.jl:
```diff
@@ -0,0 +1,294 @@
+        file["time"] = t
+        file["timestep"] = iter
+        file["point_coordinates"] = coords
+        file["values"] = values
```
```suggestion
        file["point_data"] = values
```
As a suggestion, to keep the same naming convention as for the time series output files?

DanielDoehring committed:
Update src/callbacks_step/analysis_surface_2d.jl

Co-authored-by: Michael Schlottke-Lakemper <michael@sloede.com>

Arpit-Babbar committed:
Add test

Arpit-Babbar committed:
Bug fix

Arpit-Babbar committed:
Accidental add fix

Arpit-Babbar commented:
I have added a test that validates the code by reading the final `h5` file and checking the first ten values. Is that okay?

I have also updated the [gist](https://gist.github.com/Arpit-Babbar/924823c08f4be962856a99c107b18fb5) to read `h5` files which confirms that the forces are being computed correctly.

DanielDoehring commented:
@Arpit-Babbar can you also upload the h5 files as gists (not sure if this is possible) that you are [comparing against](https://gist.github.com/Arpit-Babbar/924823c08f4be962856a99c107b18fb5)?
I would like to take a look at them, just to verify everything is working :)

DanielDoehring commented on src/callbacks_step/analysis_surface_2d.jl:
````diff
@@ -0,0 +1,294 @@
+
+Compute the surface pressure coefficient
+```math
+C_p \\coloneqq \\frac{ p - p_{p_\\infty}}
+                        {0.5 \\rho_{\\infty} U_{\\infty}^2 L_{\\infty}}
````
```suggestion
C_p \\coloneqq \\frac{p - p_{p_\\infty}}
                     {0.5 \\rho_{\\infty} U_{\\infty}^2 L_{\\infty}}
```

DanielDoehring commented on src/callbacks_step/analysis_surface_2d.jl:
````diff
@@ -0,0 +1,294 @@
+
+Compute the surface skin friction coefficient
+```math
+C_f \\coloneqq \\frac{\\boldsymbol (\\tau_w  \\boldsymbol n) \\cdot \\boldsymbol b^\\perp}
+                        {0.5 \\rho_{\\infty} U_{\\infty}^2 L_{\\infty}}
````
```suggestion
C_f \\coloneqq \\frac{\\tau_w  \\boldsymbol n \\cdot \\boldsymbol b^\\perp}
                     {0.5 \\rho_{\\infty} U_{\\infty}^2 L_{\\infty}}
```

DanielDoehring committed:
Apply suggestions from code review

DanielDoehring reviewed (dismissed):
Looking good to me! As I was involved in the coding, it would be nice to get an additional review, preferably by someone who coded up some h5 export as well, such as @sloede and/or @andrewwinters5000 

These are the plots from Arpits gist:

![c_p](https://github.com/trixi-framework/Trixi.jl/assets/75639095/16a83323-ca56-46b5-ae54-db25e057fd44)
![c_f](https://github.com/trixi-framework/Trixi.jl/assets/75639095/2128371a-1ab1-4826-b6e7-dc3f825765d2)

DanielDoehring committed:
Merge branch 'main' into surface_pressure_friction

DanielDoehring commented on examples/p4est_2d_dgsem/elixir_navierstokes_NACA0012airfoil_mach08.jl:
```diff
@@ -1,7 +1,6 @@
 using OrdinaryDiffEq
 using Trixi
-
 ###############################################################################
```
```suggestion

###############################################################################
```

DanielDoehring committed:
Update examples/p4est_2d_dgsem/elixir_navierstokes_NACA0012airfoil_mach08.jl

DanielDoehring committed:
Merge branch 'main' into surface_pressure_friction

sloede commented on src/callbacks_step/analysis.jl:
```diff
@@ -26,8 +26,8 @@ or `extra_analysis_errors = (:conservation_error,)`.
 Note: `default_analysis_errors` are `:l2_error` and `:linf_error` for all equations.
-If you want to compute `extra_analysis_errors` such as `:conservation_error` solely, i.e., 
-without `:l2_error, :linf_error` you need to specify 
+If you want to compute `extra_analysis_errors` such as `:conservation_error` solely, i.e.,
```
Can't attach this to the proper line, but the docstring should include the pointwise analysis in its list of arguments and ideally also (briefly) explain how it can be properly used.

sloede commented on src/callbacks_step/analysis_surface_integral_2d.jl:
```diff
@@ -48,31 +48,31 @@ struct AnalysisSurfaceIntegral{Variable}
 end
 
-struct ForceState{RealT <: Real}
+struct FlowStateDirectional{RealT <: Real}
```
Why the name change, or at least why the "Directional" suffix? Are their other states that might be considered? Maybe "FreestreamFlowState" would be more appropriate?

Also, what does `psi` stand for?

DanielDoehring replied:
Regarding the name change, I felt `FlowState` would be more descriptive.

`Directional` is since this flow state carries the vector of the free-stream velocity `uinf`.

The non-directional version of this has only the remaining three quantities.

sloede commented on src/callbacks_step/analysis_surface_2d.jl:
```diff
@@ -0,0 +1,294 @@
+                        values, t, iter)
+end
+
+varname(::Any) = @assert false "Surface variable name not assigned" # This makes sure default behaviour is not overwriting
```
```suggestion
```
Why define a default at all?

Arpit-Babbar replied:
This is with the intention of giving a helpful error message in case a user forgets to define `varname` for their new surface quantity.

DanielDoehring replied:
I am also not too sure about this one.

sloede commented on src/callbacks_step/analysis_surface_2d.jl:
```diff
@@ -0,0 +1,294 @@
+end
+
+varname(::Any) = @assert false "Surface variable name not assigned" # This makes sure default behaviour is not overwriting
+varname(pressure_coefficient::SurfacePressureCoefficient) = "CP_x"
+varname(friction_coefficient::SurfaceFrictionCoefficient) = "CF_x"
```
Why the spelling "CP_x" and "CF_x"? Usually I've seen these values as $c_p$ and $c_f$ in literature, so wouldn't "c_p" and "c_f" be more apt? But I am not an aerodynamics expert, so maybe I'm wrong

DanielDoehring replied:
What I am trying to denote with this is $C_P(x)$ and $C_F(x)$ to distinguish this from $C_P$ which could be understood as the integrated pressure coefficient.

sloede commented on src/callbacks_step/analysis_surface_2d.jl:
```diff
@@ -0,0 +1,294 @@
+function save_pointwise_file(output_directory, varname, coords, values, t, iter)
+    n_points = length(values)
+
+    filename = joinpath(output_directory, varname) * @sprintf("_%06d.h5", iter)
```
I suggest to start the file name with something generic, such that you can easily find all similar files, e.g., `pointwise_...` or `surface_data_` etc.

sloede reviewed (commented):
Overall, this is a nice feature addition to Trixi.jl, thanks a lot for your effort!

Generally, I am not 100% sure if the current implementation is really the optimal way to go: Naively, I assume much more code could probably be shared with the `AnalysisSurfaceIntegral` implementation. Also, I am wondering if it really makes sense to write out these rather heavy files from the `AnalysisCallback` or whether they would make more sense in the `TimeSeriesCallback`.

I've left a few more comments (some rather general ones, some more specific ones). I do not want to slow down your development pace, but on the other hand I'd like to avoid that we have a multitude of (or too many) different analysis-like tools (integral quantities over the entire domain, integral quantities over surfaces, pointwise time series over the whole domain, pointwise time series over surfaces) with vastly differing implementations, making it harder to maintain in the intermediate and long term.

Arpit-Babbar commented on src/callbacks_step/analysis_surface_2d.jl:
```diff
@@ -0,0 +1,294 @@
+            global_node_index += 1
+        end
+    end
+    # Sort with ascending x
+    sorting_indices = sortperm(coords[:, 1])
+    coords = coords[sorting_indices, :]
+    values = values[sorting_indices]
```
The sorting does mix up the top and bottom values. Yes, we should not do it.

DanielDoehring committed:
Merge branch 'surface_pressure_friction' of https://github.com/Arpit-Babbar/Trixi.jl into ArpitBabbar_PointWiseSurfForces

DanielDoehring committed:
remove sorting

DanielDoehring committed:
Update examples/p4est_2d_dgsem/elixir_navierstokes_NACA0012airfoil_mach08.jl

Co-authored-by: Michael Schlottke-Lakemper <michael@sloede.com>

DanielDoehring committed:
Update src/callbacks_step/analysis.jl

Co-authored-by: Michael Schlottke-Lakemper <michael@sloede.com>

DanielDoehring committed:
pointwise

DanielDoehring commented on src/callbacks_step/analysis_surface_2d.jl:
To be able to call this from one function we would need to add variables to the `AnalysisSurfaceIntegral/Pointswise` structs.
For Integral, this would be  `surface_integral = zero(eltype(u))` and for the pointwise stuff 
```julia
coordinates = Matrix{real(dg)}(undef, n_elements * n_nodes, dim) # physical coordinates of indices
values = Vector{real(dg)}(undef, n_elements * n_nodes) # variable values at indices
```

Then we still have to figure out how to treat the passing in of the iteration number.
Having said this, while we certainly have code-duplication in the boilerplate part, we currently have `analyze` methods with 4 different function signatures.

DanielDoehring committed:
rename

DanielDoehring committed:
viscous spelled out

DanielDoehring committed:
fmt

DanielDoehring committed:
Merge branch 'main' into surface_pressure_friction

DanielDoehring committed:
distinguish func and var

DanielDoehring committed:
fix coords

DanielDoehring committed:
fmt

DanielDoehring commented on src/callbacks_step/analysis_surface_2d.jl:
```diff
@@ -0,0 +1,294 @@
+    n_nodes = nnodes(dg)
+    n_elements = length(indices)
+
+    coords = Matrix{real(dg)}(undef, n_elements * n_nodes, dim) # physical coordinates of indices
```
Time to rename [`get_node_coords`](https://github.com/search?q=repo%3Atrixi-framework%2FTrixi.jl%20get_node_coords&type=code) ?

DanielDoehring commented:
We should put this on hold until https://github.com/trixi-framework/Trixi.jl/pull/1959 is tackled / we decide how we address https://github.com/trixi-framework/Trixi.jl/issues/1955

DanielDoehring committed:
Merge branch 'main' into surface_pressure_friction

DanielDoehring committed:
update

DanielDoehring committed:
correct docstring

DanielDoehring committed:
sort arrays

DanielDoehring committed:
mno in place

DanielDoehring commented:
@Arpit-Babbar : I did some changes to be AMR-ready, see https://github.com/trixi-framework/Trixi.jl/pull/1959 . Now the test fail, however. Can you take a look if we still match the reference data or if I messed something up?

Arpit-Babbar commented:
> Looking good to me! As I was involved in the coding, it would be nice to get an additional review, preferably by someone who coded up some h5 export as well, such as @sloede and/or @andrewwinters5000
> 
> These are the plots from Arpits gist:
> 
> ![c_p](https://private-user-images.githubusercontent.com/75639095/330315434-16a83323-ca56-46b5-ae54-db25e057fd44.svg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA0MjE4NzAsIm5iZiI6MTcyMDQyMTU3MCwicGF0aCI6Ii83NTYzOTA5NS8zMzAzMTU0MzQtMTZhODMzMjMtY2E1Ni00NmI1LWFlNTQtZGIyNWUwNTdmZDQ0LnN2Zz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA4VDA2NTI1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZhNDc5ZTNiZTJiMTRjZDY4MDZiY2M1Y2UxNDc4NTcwOTE0MWMyMjFiOGE3Nzg4YTI5OGZlNzg5ZDRlZDBmNTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.yiSMh8UWkunIdJ5Lpbn-BuSJMh6kUUOmAd8BIrKfL68) ![c_f](https://private-user-images.githubusercontent.com/75639095/330315473-2128371a-1ab1-4826-b6e7-dc3f825765d2.svg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA0MjE4NzAsIm5iZiI6MTcyMDQyMTU3MCwicGF0aCI6Ii83NTYzOTA5NS8zMzAzMTU0NzMtMjEyODM3MWEtMWFiMS00ODI2LWI2ZTctZGMzZjgyNTc2NWQyLnN2Zz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA4VDA2NTI1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE1MmFjNzVkODc4M2Y2NWE4OGVmY2E4ZTBlODJlMjdkNGE3MWUzOTllMTlkNGI0ODdmZjUyMTIyNzFiNDgyNWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XqAiM2lwFtFBVyw3Sbw1h6wIsoJ3f2gYsy19AiGAr04)

I regenerated the data and saw that both the plots look exactly the same as this. Did we remove the reordering (following Michael's suggestion?)? That will need the CI to be updated.

DanielDoehring commented:
> I regenerated the data and saw that both the plots look exactly the same as this. Did we remove the reordering (following Michael's suggestion?)? That will need the CI to be updated.

That is good!

Yes, the sorting is not longer done.
Take a look at the test files/CI test, I now sort there, but the values do not agree. This is why I asked you to check this again :)

Arpit-Babbar commented:
> Yes, the sorting is not longer done. Take a look at the test files/CI test, I now sort there, but the values do not agree. This is why I asked you to check this again :)

I see it now. That sorting does not look equivalent to sorting along the increasing x direction.
```julia
    @test sort(cp_vals[1:10]) ≈ [1.5152278762705806
           1.5182860925755697
           1.7417814818119175
           1.871513931283643
           1.874247697759559
           2.0354491762572806
           2.0400855926847936
           2.0956920203526193
           2.098639791879171
           2.1591959225932777]
```
For that, you will have to sort along `read(Trixi.h5open("out/CP_x_000007.h5"))["point_coordinates"][:,1]`

DanielDoehring commented:
> > Yes, the sorting is not longer done. Take a look at the test files/CI test, I now sort there, but the values do not agree. This is why I asked you to check this again :)
> 
> I see it now. The sorting

> does not look equivalent to sorting along the increasing x direction.

Yes, but this is also not intended. As you can see, the reference data is also sorting in ascending order, so I seek to do the same with the data, as I am not sure whether the order of the indices is deterministic.

We match some values, while some we miss, see : https://github.com/trixi-framework/Trixi.jl/actions/runs/9810227214/job/27089931612?pr=1920#step:7:8617

DanielDoehring committed:
Use counter instead of index to avoid confusion

DanielDoehring committed:
Comments

DanielDoehring committed:
Merge branch 'main' into surface_pressure_friction

DanielDoehring committed:
Update docstring

DanielDoehring committed:
More comments

DanielDoehring committed:
fmt

DanielDoehring committed:
update test vals

DanielDoehring committed:
do not pollute `analysis.dat`

DanielDoehring commented:
I guess we could also move this into a package called `TrixiAero` or something similar that implements this specialized analysis routines, thereby not polluting the `analyze` file even more.

DanielDoehring committed:
Merge branch 'main' into surface_pressure_friction
